### PR TITLE
Metadata create template 

### PR DESCRIPTION
**Pull request by joshsadam**
## Description of changes

Updated creating a metadata template to use Ant Design. Allow users to select from current metadata fields to create a new template.

## Related issue

N/A

## Checklist

Things for the developer to confirm they've done before the PR should be accepted:

*   [x] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
*   [x] Tests added (or description of how to test) for any new features.
*   [x] User documentation updated for UI or technical changes.

### CODE REVIEW

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    create one.
+     */
+    if (!loading) {
+      const found = templates.find((template) => template.identifier === id);
+
+      if (found) {
+        setTemplate(found);
+      } else if (templates.length === 0) {
+        navigate(`../fields`).then(() =>
+          notification.warn({ message: "__No templates found" })
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx:42]: Add translation

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        );
+      } else {
+        navigate(`../templates`).then(() =>
+          notification.warn({ message: "__Template cannot be found" })
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx:46]: Add translation here as well

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            />
+          </List.Item>
+          <List.Item>
+            <List.Item.Meta
+              title={<Text strong>{i18n("MetadataTemplate.fields")}</Text>}
+              description={
+                <Table
+                  pagination={false}
+                  columns={[
+                    { title: i18n("MetadataField.label"), dataIndex: "label" },
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx:75]: Move these into a const above called metadataColumns or something like that

File: src/main/webapp/resources/js/pages/projects/redux/fieldsSlice.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
+import { getMetadataFieldsForProject } from "../../../apis/metadata/field";
+import { addKeysToList } from "../../../utilities/http-utilities";
+
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/projects/redux/fieldsSlice.js:4]: File comments

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
Looking great and works very smoothly. Good job! Just some changes below

File: doc/user/user/sample-metadata/index.md [hunk 5]
```diff
@@ -134,3 +134,50 @@ Hovering over the `Export` button which will reveal a dropdown where you can sel
...
+### Metadata Templates
+
+Metadata Templates provide a pre-defined collection and order of metadata fields used for displaying sample metadata on the line list page.
+
+To create a new metadata template (**project managers only**) select the fields from the metadata fields table and click the "Create Template" button:
+
+![Video displaying how to select fields and create a new template](images/sample-metadata-create-template.gif)
+
+ * **Name**: Must be unique within the project (required)
+ * **Description**: Any information that well help to understand why this template should be used.
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply doc/user/user/sample-metadata/index.md:164]: Fixed in 745fbf77c22c8a2e8b4b9779604860392409953a

File: doc/user/user/sample-metadata/index.md [hunk 6]
```diff
@@ -134,3 +134,50 @@ Hovering over the `Export` button which will reveal a dropdown where you can sel
...
+ * **Fields**: Fields can be dragged and dropped into any order.
+
+A list of all metadata templates on the project can be seen in the "Metadata Template" menu option in the left menu:
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply doc/user/user/sample-metadata/index.md:167]: Updated in cb0955c7d85031f5837fbee8658b2b592bb95337

File: doc/user/user/sample-metadata/index.md [hunk 7]
```diff
@@ -134,3 +134,50 @@ Hovering over the `Export` button which will reveal a dropdown where you can sel
...
+  * Template actions, including (download Excel file of template fields, and remove template (managers only))
+ 
+Template Details Page
+---------------------
+
+Clicking on a template name will redirect you to the template details page.
+
+![Image displaying the metadata template details page](images/metadata-template-details.png)
+
+Project manage can edit/update the template name, description and fields on this page.
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply doc/user/user/sample-metadata/index.md:183]: Updated in 300fe679aa413ea9617bf166fecb9ceed5c8b15c

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/ajax/metadata/MetadataAjaxController.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+	@Autowired
+	public MetadataAjaxController(UIMetadataService service) {
+		this.service = service;
+	}
+
+	/**
+	 * Get a list of metadata templates for a specific project
+	 *
+	 * @param projectId Identifier for the project to get templates for.
+	 * @return List of metadata templates with associate details.
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/ajax/metadata/MetadataAjaxController.java:36]: Updated in 2470b34443694cd8b598130e40cdcf388ad96d2f

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/ajax/metadata/MetadataAjaxController.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+	/**
+	 * Updated the fields in a {@link MetadataTemplate}
+	 *
+	 * @param template the updated template to save
+	 * @param locale     Current users {@link Locale}
+	 * @return Message for UI to display about the result of the update.
+	 */
+	@PutMapping("/templates/{templateId}")
+	public ResponseEntity<AjaxResponse> updatedMetadataTemplate(@RequestBody MetadataTemplate template, Locale locale) {
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/ajax/metadata/MetadataAjaxController.java:64]: Updated name in 6f5209b90878b501d9c20603446ed3da1eb2a652

File: src/main/resources/i18n/messages.properties [hunk 10]
```diff
@@ -993,6 +993,39 @@ project.settings.coverage.maximum_coverage=Maximum Coverage:
...
+MetadataTemplatesList.delete-confirm=Delete this template?
+MetadataTemplatesList.empty=There are no metadata templates associated with this project.
+
+CreateMetadataTemplate.title=Create Metadata Template
+CreateMetadataTemplate.ok-text=Create Template
+CreateMetadataTemplate.name.required=Please provide a unique name for this template
+CreateMetadataTemplate.name.duplicate=A template with this name already exists in this project, please provide a unique name.
+
+MetadataAddTemplateField.button=Add Field
+MetadataAddTemplateField.title=Select fds to add to template
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/resources/i18n/messages.properties:1015]: Fixed in ec009a4943a559321a1beeaa19d10392e16b74c6

File: src/main/webapp/resources/js/apis/metadata/field.js [hunk 11]
```diff
@@ -15,6 +16,17 @@ const URL = setBaseUrl(`linelist/fields`);
...
   return axios({
     method: "get",
-    url: `${URL}?projectId=${projectId}`
+    url: `${OLD_URL}?projectId=${projectId}`,
   });
 }
+
+const URL = setBaseUrl(`/ajax/metadata/fields`);
+
+export async function getMetadataFieldsForProject(projectId) {
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/webapp/resources/js/apis/metadata/field.js:25]: Updated in 76bc3dc5ea84a9e35c6ed8569cb3946a19474f06

File: src/main/webapp/resources/js/apis/metadata/field.js [hunk 12]
```diff
@@ -4,9 +4,10 @@
 import axios from "axios";
 import { setBaseUrl } from "../../utilities/url-utilities";
 
-const URL = setBaseUrl(`linelist/fields`);
+const OLD_URL = setBaseUrl(`linelist/fields`);
 
 /**
+ * @deprecated
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/webapp/resources/js/apis/metadata/field.js:10]: Updated in 7901dac70ff00825a08a987ca8d26667ffeef665

File: src/main/webapp/resources/js/apis/metadata/metadata-templates.js [hunk 13]
```diff
@@ -13,3 +13,38 @@ export function getProjectMetadataTemplates(projectId) {
     .get(`${BASE_URL}?projectId=${projectId}`)
     .then(({ data }) => data);
 }
+
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/webapp/resources/js/apis/metadata/metadata-templates.js:16]: Updated in 419756a780eecff58f9d04b3e99af642dcad21ba

File: src/main/webapp/resources/js/components/ant.design/DnDTable/DraggableBodyRow.jsx [hunk 14]
```diff
@@ -0,0 +1,46 @@
+import React from "react";
+import { useDrag, useDrop } from "react-dnd";
+
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/webapp/resources/js/components/ant.design/DnDTable/DraggableBodyRow.jsx:3]: Updated in 2d552ea

File: src/main/webapp/resources/js/components/ant.design/DnDTable/index.js [hunk 15]
```diff
@@ -0,0 +1,66 @@
+import styled from "styled-components";
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/webapp/resources/js/components/ant.design/DnDTable/index.js:1]: Updated in 2d552ea

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataAddTemplateField.jsx [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+ * @constructor
+ */
+export function MetadataAddTemplateField({ fields = [], onAddFields }) {
+  const [visible, setVisible] = React.useState(false);
+  const [selected, setSelected] = React.useState([]);
+  const [selectedFields, setSelectedFields] = React.useState([]);
+
+  React.useEffect(() => {
+    /*
+    When fields are selected, Ant Table only five the key, here we are setting
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/projects/metadata/MetadataAddTemplateField.jsx]: Updated in 15aa314

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateCreate.jsx [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+  React.useEffect(() => {
+    if (fields.length) {
+      setFieldsState(addKeysToList(fields, "field"));
+    }
+  }, [fields]);
+
+  React.useEffect(() => {
+    /*
+    Need to get a list of template names so that there can only be distinct
+    names.
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateCreate.jsx]: Updated in 14ace2b

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateManager.jsx [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+          notification.warn({ message: i18n("MetadataTemplate.not-found") })
+        );
+      }
+    }
+  }, [id, templates]);
+
+  React.useEffect(() => {
+    if (Array.isArray(fields) && Array.isArray(allFields)) {
+      /*
+      Whenever the fields on the template are updated (or initial loaded), we
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateManager.jsx]: Updated in 88e2e1a

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateManager.jsx [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+      determine which fields from the project are not currently on the template.
+      This allows us to be able to only display the fields that need to be displayed
+      when the user want to add a new field to the template.
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateManager.jsx]: Updated in 8a5ec63

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateCreate.jsx [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+            size="small"
+            data={fieldsState}
+            onRowUpdate={setFieldsState}
+            columns={[
+              {
+                title: (
+                  <>
+                    <Text strong>{i18n("MetadataTemplate.fields")}</Text>
+                    <HelpPopover
+                      content={<div>{i18n("MetadataTemplateAdmin.drag")}</div>}
```
[the remark this answers is not in the document]
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateCreate.jsx]: Updated in 36f6dd7

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx (hunk 1 above)
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx]: Updated in f53975c

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx (hunk 2 above)
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx]: Updated in f53975c

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx (hunk 3 above)
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx:79]: Not worth it for 2 column declaration.  More overhead than it is worth.

File: src/main/webapp/resources/js/pages/projects/redux/fieldsSlice.js (hunk 4 above)
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/projects/redux/fieldsSlice.js:4]: Added in 88cfc10

File: src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx (hunk 3 above)
Comment (deepsidhu85) [reply src/main/webapp/resources/js/pages/projects/metadata/MetadataTemplateMember.jsx:79]: Sounds good. Thanks

**Review by deepsidhu85 — APPROVED** [review]

**Pull request merged**